### PR TITLE
GraphQL API

### PR DESCRIPTION
**Pull request by m1guelpf**
Add support for the Github GraphQL API.

### CODE REVIEW

File: lib/Github/Api/GraphQL.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?php
+namespace Github\Api;
```
Comment (cursedcoder) [inline lib/Github/Api/GraphQL.php]: 1 space line before namespace tag is required

**Comment by m1guelpf** [comment]
@cursedcoder Done!

File: lib/Github/Api/GraphQL.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Github\Api;
+
+/**
+ * GraphQL API.
+ *
+ * @link   http://developer.github.com/v3/markdown/
```
Comment (xabbuh) [inline lib/Github/Api/GraphQL.php]: Shouldn't this be https://developer.github.com/early-access/graphql/?

Comment (Nyholm) [inline lib/Github/Api/GraphQL.php]: Yes, it should

File: lib/Github/Api/GraphQL.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Github\Api;
+
+/**
+ * GraphQL API.
```
Comment (Nyholm) [inline lib/Github/Api/GraphQL.php]: Should be mention that this is the Early Access program?

Comment (m1guelpf) [inline lib/Github/Api/GraphQL.php]: Done

**Comment by m1guelpf** [comment]
@Nyholm I keep having the same problem with the tests. I don't know how the fake data is generated, so I cannot add simple GraphQL data.
Maybe you should add that to the test docs.

File: lib/Github/Api/GraphQL.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * @author Miguel Piedrafita <soy@miguelpiedrafita.com>
+ */
+class GraphQL extends AbstractApi
+{
+    /**
+     * @param string $query
+     *
+     * @return array
+     */
+    public function graphql($query)
```
Comment (Nyholm) [inline lib/Github/Api/GraphQL.php]: Do we want the signature to be: 
`$github->graphql()->graphql($query)`

Compare to: 
`$github->graphql()->execute($query)`

Comment (m1guelpf) [inline lib/Github/Api/GraphQL.php]: Do you want me to change it?

**Comment by Nyholm** [comment]
This test would work (I've not run it)

```php
class GraphQLTest extends TestCase
{
    // ...

    /**
     * @test
     */
    public function shouldShowGistComment()
    {
        $api = $this->getApiMock();

        $api->expects($this->once())
            ->method('post')
            ->with('/graphql', 'bar')
            ->will($this->returnValue('foo'));

        $result = $api-> graphql('bar');
        $this->assertEquals('foo', $result);
    }

    protected function getApiClass()
    {
        return \Github\Api\GraphQL::class;
    }
}



```

File: lib/Github/Api/GraphQL.php (hunk 4 above)
Comment (Nyholm) [inline lib/Github/Api/GraphQL.php]: What do you think? What does @cursedcoder think?

Comment (m1guelpf) [inline lib/Github/Api/GraphQL.php]: I think execute() is more clear.

**Comment by m1guelpf** [comment]
@Nyholm Added Test.

File: lib/Github/Api/GraphQL.php (hunk 4 above)
Comment (m1guelpf) [inline lib/Github/Api/GraphQL.php]: @Nyholm Changed.

**Comment by m1guelpf** [comment]
@Nyholm Changed function name.

File: test/Github/Tests/Api/GraphQLTest.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
+class GraphQLTest extends TestCase
+{
+    // ...
```
Comment (Nyholm) [inline test/Github/Tests/Api/GraphQLTest.php]: copy/pate typo

File: test/Github/Tests/Api/GraphQLTest.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
+class GraphQLTest extends TestCase
```
Comment (Nyholm) [inline test/Github/Tests/Api/GraphQLTest.php]: This file should have class comment and namespace

File: test/Github/Tests/Api/GraphQLTest.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+{
+    // ...
+
+    /**
+     * @test
+     */
+    public function shouldShowGistComment()
```
Comment (Nyholm) [inline test/Github/Tests/Api/GraphQLTest.php]: Update this method name.

File: lib/Github/Api/GraphQL.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Github\Api;
+
+/**
+ * GraphQL API.
+ *
+ * Part of the Github API Early-Access Program
+ * @link   https://developer.github.com/early-access/graphql/
```
Comment (Nyholm) [inline lib/Github/Api/GraphQL.php]: Should be an extra space before this line.

File: test/Github/Tests/Api/GraphQLTest.php (hunk 6 above)
Comment (m1guelpf) [inline test/Github/Tests/Api/GraphQLTest.php]: Done (I think)

File: test/Github/Tests/Api/GraphQLTest.php (hunk 5 above)
Comment (m1guelpf) [inline test/Github/Tests/Api/GraphQLTest.php]: What typo?

File: test/Github/Tests/Api/GraphQLTest.php (hunk 7 above)
Comment (m1guelpf) [inline test/Github/Tests/Api/GraphQLTest.php]: Done

File: lib/Github/Api/GraphQL.php (hunk 8 above)
Comment (m1guelpf) [inline lib/Github/Api/GraphQL.php]: Done

**Comment by m1guelpf** [comment]
@Nyholm How can I fix this?
ˋˋˋ
There was 1 failure:
1) Github\Tests\Api\GraphQLTest::shouldTestGraphQL
Expectation failed for method name is equal to <string:post> when invoked 1 time(s)
Parameter 1 for invocation Github\Api\AbstractApi::post('/graphql', Array (...), Array ()) does not match expected value.
Array (...) does not match expected type "string".
/home/travis/build/KnpLabs/php-github-api/lib/Github/Api/GraphQL.php:26
/home/travis/build/KnpLabs/php-github-api/test/Github/Tests/Api/GraphQLTest.php:21
ˋˋˋ

File: test/Github/Tests/Api/GraphQLTest.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Github\Tests\Api;
+
+class GraphQLTest extends TestCase
+{
+
+    /**
+     * @test
+     */
+
```
Comment (Nyholm) [inline test/Github/Tests/Api/GraphQLTest.php]: Remove this space.

File: test/Github/Tests/Api/GraphQLTest.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    public function shouldTestGraphQL()
+    {
+        $api = $this->getApiMock();
+
+        $api->expects($this->once())
+            ->method('post')
+            ->with('/graphql', 'bar')
```
Comment (Nyholm) [inline test/Github/Tests/Api/GraphQLTest.php]: Sorry, I missed that you are building an array. 
```php
->with($this->equalTo('/graphql'), $this->equalTo(['query'=>'bar']))
```

**Comment by m1guelpf** [comment]
@Nyholm Fixed.

**Comment by m1guelpf** [comment]
@Nyholm @cursedcoder Merge?

**Comment by Nyholm** [comment]
Im 👍  to merge. But since other collaborators were involved here I would like to give them some time to review the changes as well.

**Comment by cursedcoder** [comment]
@xabbuh what would you say?

**Comment by cursedcoder** [comment]
Anyway merging this because we got 2/3 votes.

**Pull request merged**